br/pkg/restore/restorer.go

```diff
@@ -0,0 +1,348 @@
+	start := time.Now()
+	fileCount := 0
+	defer func() {
+		elapsed := time.Since(start)
+		if err == nil {
+			log.Info("Restore files", zap.Duration("take", elapsed))
+			summary.CollectSuccessUnit("files", fileCount, elapsed)
+		}
+	}()
+
+	log.Debug("start to restore files", zap.Int("files", fileCount))
+
+	if span := opentracing.SpanFromContext(m.ectx); span != nil && span.Tracer() != nil {
+		span1 := span.Tracer().StartSpan("Client.RestoreSSTFiles", opentracing.ChildOf(span.Context()))
+		defer span1.Finish()
+		m.ectx = opentracing.ContextWithSpan(m.ectx, span1)
+	}
+
+	for _, batchFileSet := range batchFileSets {
+		if m.ectx.Err() != nil {
+			log.Warn("Restoring encountered error and already stopped, give up remained files.",
+				logutil.ShortError(m.ectx.Err()))
+			// We will fetch the error from the errgroup then (If there were).
+			// Also note if the parent context has been canceled or something,
+			// breaking here directly is also a reasonable behavior.
+			break
+		}
+		filesReplica := batchFileSet
+		m.fileImporter.WaitUntilUnblock()
+		m.workerPool.ApplyOnErrorGroup(m.eg, func() (restoreErr error) {
+			fileStart := time.Now()
+			defer func() {
+				if restoreErr == nil {
+					log.Info("import files done", zap.Duration("take", time.Since(fileStart)))
+					onProgress(int64(len(filesReplica)))
+				}
+			}()
+			if importErr := m.fileImporter.Import(m.ectx, filesReplica...); importErr != nil {
+				return errors.Trace(importErr)
+			}
+
+			// the data of this range has been import done
+			if m.checkpointRunner != nil && len(filesReplica) > 0 {
+				for _, filesGroup := range filesReplica {
+					rangeKeySet := make(map[string]struct{})
+					for _, file := range filesGroup.SSTFiles {
+						rangeKey := GetFileRangeKey(file.Name)
+						// Assert that the files having the same rangeKey are all in the current filesGroup.Files
+						rangeKeySet[rangeKey] = struct{}{}
+					}
+					for rangeKey := range rangeKeySet {
+						// The checkpoint range shows this ranges of kvs has been restored into
+						// the table corresponding to the table-id.
+						if err := checkpoint.AppendRangesForRestore(m.ectx, m.checkpointRunner, filesGroup.TableID, rangeKey); err != nil {
+							return errors.Trace(err)
+						}
+					}
+				}
+			}
+			return nil
+		})
+	}
+	// Once the parent context canceled and there is no task running in the errgroup,
+	// we may break the for loop without error in the errgroup. (Will this happen?)
+	// At that time, return the error in the context here.
+	return m.ectx.Err()
+}
+
+func GetFileRangeKey(f string) string {
+	// the backup date file pattern is `{store_id}_{region_id}_{epoch_version}_{key}_{ts}_{cf}.sst`
+	// so we need to compare with out the `_{cf}.sst` suffix
+	idx := strings.LastIndex(f, "_")
+	if idx < 0 {
+		panic(fmt.Sprintf("invalid backup data file name: '%s'", f))
+	}
+
+	return f[:idx]
+}
+
+type PipelineRestorerWrapper[T any] struct {
```

Comment:
Perhaps this can be a plain function. So the type argument can be deduced, no explicit annotations like `restore.PipelineRestorerWrapper[*logclient.LogDataFileInfo]` needed.